PWGMM/Tasks/run2dndeta.cxx

```diff
@@ -100,23 +114,55 @@ struct PseudorapidityDensity {
   void process(soa::Join<aod::Collisions, aod::EvSels>::iterator const& collision, Trks const& tracks)
   {
     registry.fill(HIST("EventSelection"), 1.);
-    if (!useEvSel || (useEvSel && collision.sel7())) {
-      registry.fill(HIST("EventSelection"), 2.);
-      auto z = collision.posZ();
-      auto perCollisionSample = sample->sliceByCached(aod::track::collisionId, collision.globalIndex());
-      if (perCollisionSample.size() > 0) {
-        registry.fill(HIST("EventSelection"), 3.);
+    if (doprocessGen) //we are sudying the MC, that needs correction
+    {
+      if (!useEvSel || (useEvSel && collision.sel7())) {
+
+        auto z = collision.posZ();
+        auto perCollisionSample = sample->sliceByCached(aod::track::collisionId, collision.globalIndex());
+        auto v = fRatio->Eval(z);
+
+        double r = rand->Rndm(); //random number between 0 and 1
+        if ((1 - v) <= r)        //then we accept the event
```

Comment:
This way you duplicate the processing code, if you use `if (r < v) { return; }` you can just skip the collision and have the processing fragment only once after the condition.